apis/apps/v1alpha1/apicast_types.go

```diff
@@ -28,6 +30,21 @@ import (
 // EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
+// CustomPolicySpec contains or has reference to an APIcast custom policy
+type CustomPolicySpec struct {
```

Comment:
If in the future we think we might implement other ways to provide the policy maybe it would make sense to have a section for it like:

customPolicySpec:
  name:
  version:
  policySource:
    secret:
      secretRef:
        ...
    inline:
      relevantfieldToInlinePolicy1
      relevantfieldToInlinePolicy2
      ...
    ...
```

What do you think?

Reply:
`Name` and `Version` are mandatory in any case. 

I agree, The secret (v1.LocalObjectReference type) can be optional (but make it mandatory in this implementation) and let in the future add some `inline` object.

Reply:
updated

Reply:
:+1: .

Do you think it might end up being confusing if we end up having multiple common attributes together with the different policySource types (a union) at the same level? imagine there could be other unions at the same level too.

The current implementation would end up looking as:
```
customPolicySpec:
  name:
  version:
  commonAttr1:
  commonAttr2:
  secretRef:
  inline:
    ...
  anotherPolicySourceType:
    ...
  aNewUnionTypeType1:
    ...
  aNewUnionTypeType2:
    ...
  ...
```

vs what it was proposed in the initial comment:

```
customPolicySpec:
  name:
  version:
  commonAttr1:
  commonAttr2:
  policySource:
    secret:
      secretRef:
      ...
    inline:
      relevantfieldToInlinePolicy1:
      relevantfieldToInlinePolicy2:
      ...
    ...
  aNewUnionType:
    aNewUnionTypeType1:
      ...
    anewUnionTypeType2:
```

The intention of the proposal in my previous comment was to try to avoid this kind of cluttering at the same level of unrelated terms and excluding values, being contained in their own section.

Reply:
I do not think it is necessary or confusing.

There can be other objects at the same level, some of them being policy source types and others common attributes.

Reply:
> There can be other objects at the same level, some of them being policy source types and others common attributes.

I agree. But notice the example I pointed where there would be another equivalent of policy source at the same level.

Reply:
Sorry, I do not get what would be mutually exclusive. 

aNewUnionTypeType1` is like `commonAttr1` but being an object instead of a simple value. 

```yaml
customPolicySpec:
  - name:
    version:
    commonAttr1:
    commonAttr2:
    secretRef:
    inline:
      ...
    anotherPolicySourceType:
      ...
    aNewUnionTypeType1:
      ...
    aNewUnionTypeType2:
      ...
```

Reply:
> Sorry, I do not get what would be mutually exclusive.

Sorry. Let me try to clarify. An example of a "union type", where by definition all of its values are mutually exclusive would be the "policy source". We would have `secret` or `inline`. The user can't / shouldn't not set the two because they are mutually exclusive. They are part of an imaginary "union type". It is also explained a little bit in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions

Reply:
This is a documentation and internal validation issue then. 

When we add the inline policy source, if we ever add it, we need to document that it is mutually exclusive with the secret source and validate in the controller. 

I still think it is ok to go like it is now.